sui_core/src/consensus_adapter.rs

```diff
@@ -129,91 +129,91 @@ impl ConsensusAdapter {
-    async fn process_consensus_output(&mut self, output: ConsensusOutput) {
-        let (result, serialized) = output;
-
-        // Execute the transaction (if the consensus successfully sequenced it).
-        let outcome = match result {
-            Ok(()) => {
-                let message =
-                    bincode::deserialize(&serialized).expect("Failed to deserialize consensus tx");
-                let ConsensusTransaction::UserTransaction(certificate) = message;
-                let confirmation_transaction = ConfirmationTransaction { certificate };
-                self.state
-                    .handle_confirmation_transaction(confirmation_transaction)
-                    .await
-            }
-            Err(e) => Err(SuiError::from(e)),
-        };
-
-        // Notify the caller that the transaction has been sequenced (if there is a caller).
-        let digest = Self::hash(&serialized);
-        if let Some(repliers) = self.pending.remove(&digest) {
-            for replier in repliers {
-                if replier.send(outcome.clone()).is_err() {
-                    debug!("No replier to listen to consensus output {digest}");
-                }
-            }
-        }
-    }
-
     /// Main loop receiving messages input to consensus and notifying the caller once the inputs
     /// are sequenced (of if an error happened).
     async fn run(&mut self) {
         loop {
             tokio::select! {
-                // Keep track of this certificates so we can notify the user later.
-                Some(consensus_input) = self.rx_consensus_input.recv() => {
-                    let serialized = consensus_input.serialized;
-                    let replier = consensus_input.replier;
-                    let digest = Self::hash(&serialized);
-                    self.pending.entry(digest).or_insert_with(Vec::new).push(replier);
+                // A new transaction has been sent to consensus or is no longer needed.
+                Some(message) = self.rx_consensus_input.recv() => {
+                    match message {
+                        // Keep track of this certificates so we can notify the user later.
+                        ConsensusListenerMessage::New(transaction, replier) => {
+                            if self.pending.len() < self.max_pending_transactions {
+                                let digest = Self::hash(&transaction);
+                                self.pending.entry(digest).or_insert_with(Vec::new).push(replier);
+                            }
+                        },
+
+                        // Stop waiting for a consensus transaction.
+                        ConsensusListenerMessage::Cleanup(transaction) => {
+                            let digest = Self::hash(&transaction);
+                            let _ = self.pending.get_mut(&digest).and_then(|x| x.pop());
+                            if self.pending.get(&digest).map_or_else(|| false, |x| x.is_empty()) {
```

Comment:
I can't wait for `inspect` to stabilize.

Reply:
We could but it's extra LOC